### DISCUSSION TITLE
shallowProxy - similar to vue's shallowRef

### DISCUSSION THREAD

**Discussion by dts** [pmndrs/valtio#953] [Ideas]
It would be nice to be able to create a shallow proxy, something that only does a reactive transform to the first level. Vue's "shallowRef" (https://vuejs.org/api/reactivity-advanced.html#shallowref) has this behavior - and it makes it easier to "ingest" large/weird objects. Sometimes I only care about dependency tracking at the top level, because the rest of it is tracked/managed elsewhere. I think this would give a good path forward in cases similar to https://github.com/pmndrs/valtio/discussions/782.

**Comment by dai-shi** [comment]
Our solution is `ref()` in Valtio. It requires you to mark all properties you want ignore about tracking, but does it work for you?

**Reply by dts** [reply]
This has been sufficient for my case, for what it's worth.

**Reply by dai-shi** [reply]
It's not too surprising, but I would like to understand the problem and the limitation better. Can you share some code snippets to elaborate?

**Reply by dts** [reply]
Using File objects inside a valtio store. Here's a snippet of code that enforces this rule:
```export const createFileUploadStore = () => {
  const p = proxy(new FileUploadStore());

  subscribe(p, () => {
    p.files.some((f) => {
      if (getVersion(f)) {
        const error = new Error("Do not add Files directly, use ref(), addFiles, or setFiles.");

        if (p.attached) p.error ??= error;
        else throw error;
      }
    });
  });

  return p;
};
```

**Reply by dai-shi** [reply]
Interesting. Thanks for sharing!

**Comment by megazazik** [comment]
Hello. In our case, shallowProxy would be very useful. We use objects where one field should be tracked, and the rest of the fields should not. And new fields are periodically added. We have to constantly make sure that all new fields are wrapped in a ref. It is inconvenient, and it could be easy forgotten.

**Comment by dai-shi** [comment]
@overthemike Any thougths?

**Reply by overthemike** [reply]
Sorry for the delayed response. Been getting over a nasty cold.

I have some ideas on how it could be implemented via a wrapper library, but I'm not sure if I'm understanding the desired outcome exactly. Is the request for having an option to only track the top level object or is it more only tracking specified properties? Are we supposed to be ignoring new properties added to the tracked object as well? I may be misunderstanding, but I feel like there are a few different requests in this thread...

**Reply by dai-shi** [reply]
Me neither.

> I feel like there are a few different requests in this thread...

Yeah, probably.

@megazazik Would you like to elaborate your requirement?

**Reply by megazazik** [reply]
My request was for an implementation where all fields are tracked only at the top level (all these fields can be wrapped in the ref function), and all new properties also become tracked only at the top level. 
But if I manually wrap some property in the proxy, then changes in this property should be tracked in all levels.

**Reply by dai-shi** [reply]
```js
const p = proxy({});
const addPropToP = (key, val) => {
  p[key] = ref(val);
};
```

Does it solve your case, and what's missing is just a nice syntax?

**Reply by megazazik** [reply]
This example does not take into account changing and adding fields. In these cases, the new values ​​must also be wrapped in the ref.

About syntax - our code may look like this. I am not sure we could use function like `addPropToP`
```js
class A { 
  propA1 =  proxy({});
  propA2 = ref({});
}

class B extends A {
  propB1 = proxy({});
  propB2 = ref({});
}

const b = proxy(new B());
```

**Reply by dai-shi** [reply]
My question was if `ref` could technically solve you problem.

Anyway, the challenge that Valtio solves is deep proxy and deep snapshot. If the requirement is shallow proxy, it should be much easier.

```js
const SNAPSHOT = Symbol();
const LISTENERS = Symbol();

export const proxy = (baseObj) => {
  let snap = { ...baseObj };
  const listeners = new Set();
  return new Proxy(baseObj, {
    set(target, prop, value, receiver) {
      const result = Reflect.set(target, prop, value, receiver);
      snap = { ...target };
      listeners.forEach((fn) => fn());
      return result;
    },
    get(target, prop, receiver) {
      if (prop === SNAPSHOT) {
        return snap;
      }
      if (prop === LISTENERS) {
        return listeners;
      }
      return Reflect.get(target, prop, receiver);
    },
  });
};

export const snapshot = (p) => p[SNAPSHOT];

export const subscribe = (p, fn) => {
  const listeners = p[LISTENERS];
  listeners.add(fn);
  return () => listeners.delete(fn);
};

import { useCallback, useSyncExternalStore } from 'react'

export const useSnapshot = (p) => useSyncExternalStore(
  useCallback((fn) => subscribe(p, fn), [p]),
  () => snaphost(p),
);
```

**Reply by megazazik** [reply]
> My question was if `ref` could technically solve you problem.

yes, it could
we solve the problem via `ref` now, but we have to manually wrap each field in `ref`.

is your example - how we could implement swallowProxy outside valtio? or what implementation inside valtio could look like?

**Reply by dai-shi** [reply]
> but we have to manually wrap each field in ref.

Good to hear. That's what I mean by missing nice syntax.

I think double proxy should be the easiest userland solution.
```js
import { proxy, ref } from 'valtio';

export const shallowProxy = (baseObj) => {
  Object.values(baseObj).forEach((value) => ref(value));
  const p = proxy(baseObj);
  return new Proxy(p, {
    set(target, prop, value, receiver) {
      ref(value);
      return Reflect(target, prop, value, receiver);
    },
  });
};
```

**Reply by megazazik** [reply]
ok, thank you.
I will try it.

**Comment by bisubus** [comment]
I don't want to to hijack, it seems it's more about Vue's shallowReactive in general and not specifically shallowRef.

I stumbled upon this topic while trying to do the thing that the title suggests and also implement [Vue unwrapping ](https://vuejs.org/guide/essentials/reactivity-fundamentals.html#additional-ref-unwrapping-details). I don't think doing this automagically was a good move any way because it causes too much surprises, explicit unwrapping is easier to implement including TS. 

Btw Valtio's `ref` is similar to Vue's `markRaw` and not refs themselves, the naming can cause confusion in this context.

I'm trying to make proxies work with react-like refs, which seems reasonable for the integration. I've tried to hook into `proxy` through internal api instead of a double proxy:

```ts
import { type RefObject, useRef } from 'react';
import { proxy, ref as raw, unstable_replaceInternalFunction } from 'valtio';

const UNWRAP = Symbol.for('unwrap');

function unwrap<T>(value: RefObject<T>): T {
  return Object.assign(value, { [UNWRAP]: true }) as T;
}

unstable_replaceInternalFunction('canProxy', (prevFn) => {
  return (value: any) => {
    if (value?.[UNWRAP]) {
      return false;
    }

    return prevFn(value);
  };
});

unstable_replaceInternalFunction('createHandler', (prevFn) => {
  return (...args) => {
    const handler = prevFn(...args);

    handler.get = (target, prop, receiver) => {
      const value = Reflect.get(target, prop, receiver)

      return value?.[UNWRAP] ? value.current : value;
    }

    return handler;
  };
});
```

```ts
const wrappedRawRef = useRef({ a: { b: 1 } });
const rawRef = useRef({ a: { b: 1 } });
const shallowRef = useRef(raw(({ a: { b: 1 } })));
const deepRef = useRef({ a: { b: 1 } });

const state = useRef(proxy(
  {
    wrappedRawRef: raw(wrappedRawRef),
    rawRef: unwrap(raw(rawRef)),
    shallowRef: unwrap(shallowRef),
    deepRef: unwrap(proxy(deepRef))
  })).current;
```

It appears to work as expected so far:

<details>

![Screenshot](https://github.com/user-attachments/assets/d9619afe-e758-4582-ac31-ee124f674642) 

![Screenshot](https://github.com/user-attachments/assets/a6fbb412-8f92-4572-a56a-a3d0e1e40741)
</details>

@dai-shi What do you think about this approach? It seems that get trap is a good place to do this

I'm hesitating about unstable_replaceInternalFunction, and it doesn't look right to me to modify a global for local purpose. Are there plans to make these hooks public, could proxy() and other core api accept options for customization?

I've seen valtio-plugin. It seems to be generally aimed at higher level ops, what's the status of it?

**Reply by overthemike** [reply]
@bisubus `valtio-plugin` basically has 2 main goals: 1. The ability to create proxy "instances" that have their own context so that you can apply plugins to different parts of state without interfering with other parts of state; and 2. Make it easier and more approachable to hook into valtio state lifecycles instead of having to use `unstable_replaceInternalFunction` to get custom functionality. 

As far as the status of it, it's pretty much brand new and more or less experimental. I highly encourage giving it a shot. All feedback would be extremely welcome.

**Reply by bisubus** [reply]
Any suggestions how it could be used in this case? I see that onGet is a listener and I can't affect the result of a getter. And there seems to be no way to apply unstable_replaceInternalFunction to specific proxy instance either

**Reply by overthemike** [reply]
@bisubus I published a new minor version of `valtio-plugin` (1.9.0) that will allow for overriding a get call. Something like this should hopefully work. Would love to hear how it goes.

```ts
import { proxy } from 'valtio-plugin'

const UNWRAP = Symbol.for('unwrap');

const plugin: ValtioPlugin = {
  id: 'unwrap',
  onGet: (path, value) => {
    if (value?.[UNWRAP]) {
      return value.current
    }
    // if no value is returned (or undefined is returned),
    // it will return the original value so you don't need to return it here
  }
}

// create an instance - this is optional as you can also just apply this globally with proxy.use()
const proxyInstance = proxy.createInstance()
proxyInstance.use(plugin)
```

**Reply by overthemike** [reply]
There's no way currently, however, to replace the need for replacing the `canProxy` function. I'll look into that as well but you could use a hybrid approach for the moment and continue to use `unstable_replaceInternalFunction`

**Reply by bisubus** [reply]
@overthemike I appreciate the help. "if no value is returned (or undefined is returned)" - oh, this looks like an invitation for bugs, IME plain hooks/listeners need to be distinguished from transformers to avoid uncertainty like this one. Probably a pair of transformGet/transformSet could cover a lot of cases that previously required double proxies.

Here it seems like canProxy could be replaced with set transformer that wraps the value with raw(), At least if it's done the blunt way I did by adding UNWRAP to an object (though this worked for vue reactivity) instead of less obtrusive way that canProxy itself works (comparing values with a weak map)

**Reply by overthemike** [reply]
@bisubus That's a great idea. I'm currently working on an implementation for canProxy so that it can be applied to instances rather than globally overwriting the canProxy function. While I'm at it, I'll take your suggestion as well. Much appreciated.

**Reply by overthemike** [reply]
@bisubus I've pushed up a new version of `valtio-plugin` (1.10.2) that now also handles scoped `canProxy` as well as adding in `transformGet` and `transformSet` (and making onGet return void again). You can now do something like this:

```ts
import { proxy } from 'valtio-plugin'

const UNWRAP = Symbol.for('unwrap');

const plugin = {
  id: 'unwrap',
  transformGet: (path, value) => {
    if (value?.[UNWRAP]) {
      return value.current
    }
  },
  canProxy: (value, prevFn) => {
    if (value?.[UNWRAP]) {
      return false;
    }
    return prevFn(value)
  }
}

// create an instance - this is optional as you can also just apply this globally with proxy.use()
const proxyInstance = proxy.createInstance()
proxyInstance.use(plugin)
```
or you can use the `transformSet` method like you mentioned earlier to accomplish basically the same thing. Let me know if this works out.

**Reply by bisubus** [reply]
@overthemike Thanks, this looks usable. I was thinking about something with (target, prop, value, receiver, originalFn) signature for transformSet, but this one will probably work better for most cases. 

I'm not sure what were the original cases for onGet that made it appear but probably it [needs to go after transformGet](https://github.com/valtiojs/valtio-plugin/blob/main/src/index.ts#L515-L540) to receive the final value

**Reply by overthemike** [reply]
@bisubus The `onGet` is just a hook into when the user calls a property. I put it before transformGet because I would want it receive the original value before it can be transformed. You can just use transformGet otherwise. Right?

If you have anymore insights for it, I've gone ahead and opened up the discussion tab on the `valtio-plugin` repo. Let's move anymore discussions over there so we don't pollute this page. I'm really wanting to hunt down some feedback on how it's working and also always looking for contributions!

**Reply by bisubus** [reply]
@overthemike I was thinking the opposite way about it, onGet can access the final value because it cannot affect it, and intermediate value can be accessed in transformGet if needed. Not that it matters much. I'll provide more feedback as I put it into action
